fix[sml]: reduce update latencies in sml component

unaiur opened the pull request:
# What does this implement/fix?

SML component consumes too much memory, making it impossible to run in ESP8266 boards and also issues constant warnings about high latencies:

```
[17:13:31][W][component:237]: Component sml took a long time for an operation (87 ms).
[17:13:31][W][component:238]: Components should block for at most 30 ms.
```

First improvement was replacing std::vector with a new Span class inspired in C++20 std::span, removing
associated memory allocations, reducing memory requirements from ~6KiB to ~2.5KiB and reducing latencies in a ESP32-S2 from 120-130ms to 50-60ms when DEBUG logs are enabled, and from 80-90ms to less than 30ms when not. Once all frameworks use C++20, we can remove the Span class and use std::span instead.

Then, I rewrote completely our SML message parser, dropping the intermediate tree representation and extracting OBIS counters directly from  SML binary message in a single pass. This reduced drastically memory requirements to 300 bytes.

This huge memory reduction makes ESP8266 a viable target for this component and fixes https://github.com/esphome/issues/issues/4974.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**
- fixes: https://github.com/esphome/issues/issues/4974
- related to https://github.com/esphome/esphome/issues/9517


**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

Not needed.

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

unaiur commented:
Hi @alengwenus, I have been running these changes for 3 days now and the board did not rebooted  even once; before it was rebooting every 12h-24h. I think that blocking the main loop for more than 0.1 seconds causes instabilities.

Can you review my changes, please?

sur5r commented on esphome/components/sml/sml.cpp:
```diff
@@ -66,32 +66,13 @@ void Sml::add_on_data_callback(std::function<void(std::vector<uint8_t>, bool)> &
-  this->log_obis_info_(obis_info);
-}
-
-void Sml::log_obis_info_(const std::vector<ObisInfo> &obis_info_vec) {
-  ESP_LOGD(TAG, "OBIS info:");
-  for (auto const &obis_info : obis_info_vec) {
-    std::string info;
-    info += "  (" + bytes_repr(obis_info.server_id) + ") ";
-    info += obis_info.code_repr();
-    info += " [0x" + bytes_repr(obis_info.value) + "]";
-    ESP_LOGD(TAG, "%s", info.c_str());
-  }
-}
```
Why is the debug functionality being removed in this PR?

UnaiUribarri-TomTom replied:
It is not removed but refactored in the lines 74-75 (inside Sml::publish_value_), to avoid calculating the text if logging is disabled.

unaiur replied:
It is not removed but refactored in the lines 71-72 (inside Sml::publish_value_), to avoid calculating the text if logging is disabled.

zivillian commented:
I successfully tested this PR on a ESP8266 and can confirm that it works with at least 9 Sensors defined. Without this PR it crashed instantly.

>Component sml took a long time for an operation (67 ms).
Components should block for at most 30 ms.

UnaiUribarri-TomTom commented:
> I successfully tested this PR on a ESP8266 and can confirm that it works with at least 9 Sensors defined. Without this PR it crashed instantly.
> 
> > Component sml took a long time for an operation (67 ms).
> > Components should block for at most 30 ms.
> 
> 15 sensors also work, but loading the web ui causes a reboot.

Both SML and WebUI requires quite amount of memory. Maybe disabling webui and using HomeAssistant or MQTT allows you to have all 15 sensors.

unaiur committed:
fix[sml]: reduce update latencies in sml component

SML component is issuing constant warnings about high latencies:

```
[17:13:31][W][component:237]: Component sml took a long time for an operation (87 ms).
[17:13:31][W][component:238]: Components should block for at most 30 ms.
```

This change replaces std::vector with a new Span class inspired in C++20 std::span, removing
associated memory allocations and reducing latencies in a ESP32-S2 from 120-130ms to
50-60ms when DEBUG logs are enabled, and from 80-90ms to less than 30ms when not.

Once all frameworks use C++20, we can remove the Span class and use std::span instead.

unaiur committed:
fix[sml]: reduce memory requirements for SML sensor

Remove dynamic memory required to decode a SML message into its
semantic tree and replace it with a custom parser that directly
parses the binary SML message.

This change reduces RAM memory usage from 5KiB to 300 bytes and
allows to use this component in ESP8266 boards.

unaiur commented:
> I successfully tested this PR on a ESP8266 and can confirm that it works with at least 9 Sensors defined. Without this PR it crashed instantly.
> 15 sensors also work, but loading the web ui causes a reboot.

I am working on a complete rewrite of the SML parser to avoid deserializing the SML message into a semantic tree and instead directly parse the binary SML message.

Before all this changes, deserializing SML message required more than 5 KiB of RAM with a large number of memory allocations / deallocations per second.

Currently, it optimizes the tree to use less than 3KiB of RAM and reduces memory allocations / deallocations per second.

A new binary parser will completely remove the tree, only requiring dynamic memory to store the received SML binary message (around 300 bytes).

That will allow you to run SML component in ESP8266 with all 15 sensors.
